pkg/controller/policyrecommendation_controller.go

```diff
@@ -18,59 +18,194 @@ package controller
 //+kubebuilder:rbac:groups=ottoscaler.io,resources=policyrecommendations/finalizers,verbs=update
 
-// Reconcile is part of the main kubernetes reconciliation loop which aims to
-// move the current state of the cluster closer to the desired state.
-// TODO(user): Modify the Reconcile function to compare the state specified by
-// the PolicyRecommendation object against the actual cluster state, and then
-// perform operations to make the cluster state reflect the state specified by
-// the user.
-//
-// For more details, check Reconcile and its Result here:
-// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.14.1/pkg/reconcile
 func (r *PolicyRecommendationReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	logger := ctrl.LoggerFrom(ctx).WithName(POLICY_RECO_WORKFLOW_CTRL_NAME)
 
-	// TODO(user): your logic here
+	policyreco := v1alpha1.PolicyRecommendation{}
+	if err := r.Get(ctx, req.NamespacedName, &policyreco); err != nil {
+		// we'll ignore not-found errors, since they can't be fixed by an immediate
+		// requeue (we'll need to wait for a new notification), and we can get them
+		// on deleted requests.
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	logger.V(2).Info("PolicyRecomemndation retrieved", "policyreco", policyreco)
 
+	r.Recorder.Event(&policyreco, EVENT_TYPE_NORMAL, "HPARecoQueuedForExecution", "This workload has been queued for a fresh HPA recommendation.")
+
+	hpaConfigToBeApplied, targetHPAReco, policy, err := r.RecoWorkflow.Execute(ctx, reco.WorkloadMeta{
+		TypeMeta:  policyreco.Spec.WorkloadMeta.TypeMeta,
+		Name:      policyreco.Spec.WorkloadMeta.Name,
+		Namespace: policyreco.Spec.WorkloadMeta.Namespace,
+	})
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	var policyName string
+
+	if policy != nil {
+		policyName = policy.Name
+	}
+
+	transitionedAt := retrieveTransitionTime(hpaConfigToBeApplied, policyreco)
+	now := metav1.Now()
+	policyRecoPatch := &v1alpha1.PolicyRecommendation{
+		TypeMeta: policyreco.TypeMeta,
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      policyreco.Name,
+			Namespace: policyreco.Namespace,
+		},
+		Spec: v1alpha1.PolicyRecommendationSpec{
+			QueuedForExecution:      &FALSE_BOOL,
+			TargetHPAConfiguration:  *targetHPAReco,
+			Policy:                  policyName,
+			CurrentHPAConfiguration: *hpaConfigToBeApplied,
+			TransitionedAt:          &transitionedAt,
+			GeneratedAt:             &now,
+		},
+	}
+	logger.V(0).Info("Policy Patch", "PolicyReco", *policyRecoPatch)
+	if err := r.Patch(ctx, policyRecoPatch, client.Apply, client.ForceOwnership, client.FieldOwner(POLICY_RECO_WORKFLOW_CTRL_NAME)); err != nil {
+		logger.Error(err, "Error patching the policy reco object")
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+	logger.V(0).Info("Policy Patch", "PolicyReco", *policyRecoPatch)
+
+	statusPatch := &v1alpha1.PolicyRecommendation{
```

Comment:
Why are we pathching the PolicyRecommendation CR in 2 calls?

Reply:
Subresources must be updated separately from the Spec. 

https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#status-subresource
https://book-v1.book.kubebuilder.io/basics/status_subresource.html